cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/submit/AbstractSparkSubmitter.java

```diff
@@ -216,14 +218,61 @@ protected void submit(SparkRuntimeContext runtimeContext,
 
   private abstract static class SparkJobFuture<V> extends AbstractFuture<V> {
 
+    private static final Logger LOG = LoggerFactory.getLogger(SparkJobFuture.class);
+    private final AtomicBoolean done;
+    private final SparkRuntimeContext context;
+
+    protected SparkJobFuture(SparkRuntimeContext context) {
+      this.done = new AtomicBoolean();
+      this.context = context;
+    }
+
     @Override
-    public boolean set(@Nullable V value) {
-      return super.set(value);
+    protected boolean set(V value) {
+      if (done.compareAndSet(false, true)) {
+        return super.set(value);
+      }
+      return false;
     }
 
     @Override
-    public boolean setException(Throwable throwable) {
-      return super.setException(throwable);
+    protected boolean setException(Throwable throwable) {
+      if (done.compareAndSet(false, true)) {
+        return super.setException(throwable);
```

Comment:
Same as above. This is the only method in `AbstractFuture` for setting an exception to the future.